torch/csrc/jit/codegen/cuda/lower_loops.cpp

```diff
@@ -90,100 +90,154 @@ void LoopNestGenerator::pushFront(kir::Expr* expr) {
-    openFor(*loop_structure_it);
-  }
+    TORCH_INTERNAL_ASSERT(
+        n_loops_to_close >= 0 && n_loops_to_close <= for_loops_.size(),
+        "Tried to close an invalid number of loops: ",
+        n_loops_to_close);
+
+    if (min_close < n_loops_to_close && min_close > 0) {
+      // Figure out where the last for loop matches from out_tv, go until the
+      // min_close loop marked from previous tv's producer domain. Make sure
+      // none of these domains are actually present in current out_tv. If these
+      // loops map to current out_tv, it should be responsible for deciding if
+      // they stay or go, this could result from an invalid compute at topology
+      // on the DAG or bad expression sorting.
+      auto for_loops_it = for_loops_.end() - n_loops_to_close;
+      auto for_loops_it_end = for_loops_.end() - min_close;
+
+      for (; for_loops_it != for_loops_it_end; for_loops_it++) {
+        TORCH_INTERNAL_ASSERT(
+            std::none_of(
+                loop_structure_it,
+                loop_structure.end(),
+                [&gpu_lower, &for_loops_it](IterDomain* loop_structure_id) {
+                  // Check loop structure doesn't map for_loops in for loop map
+                  auto id0 = (*for_loops_it)->iter_domain();
+                  auto id1 = gpu_lower->lowerValue(loop_structure_id)
+                                 ->as<kir::IterDomain>();
+                  return gpu_lower->caLoopMap().areMapped(id0, id1) ||
+                      gpu_lower->caParallelMap().areMapped(id0, id1);
+                }),
+            "Invalid loop found to close.");
+      }
+
+      n_loops_to_close = std::min(n_loops_to_close, min_close);
+    }
 
-  pushFront(gpu_lower->lowerExpr(expr));
-}
+    for (int64_t i_loop_close = 0; i_loop_close < n_loops_to_close;
+         i_loop_close++) {
+      closeFor();
+    }
 
-// Generate the loop nest structure and place it in lowered_exprs_
-void LoopNestGenerator::generate(const std::vector<Expr*>& exprs) {
-  TORCH_INTERNAL_ASSERT(lowered_exprs_.empty());
+    // Open the remaining needed loops
+    for (; loop_structure_it != loop_structure.end(); ++loop_structure_it) {
+      openFor(*loop_structure_it);
+    }
 
-  // Process the carefully ordered expressions
-  for (auto it = exprs.rbegin(); it != exprs.rend(); ++it) {
-    handle(*it);
+    if (out_tv->getMaxProducerPosition() == 0) {
+      min_close = -1;
+    } else {
+      auto produce_at_id = loop_structure[out_tv->getMaxProducerPosition() - 1];
+      auto min_close_loop = std::find_if(
+          for_loops_.begin(),
+          for_loops_.end(),
+          [&produce_at_id, &gpu_lower](kir::ForLoop* fl) {
+            auto produce_at_lowered_it =
+                gpu_lower->lowerValue(produce_at_id)->as<kir::IterDomain>();
+            return produce_at_lowered_it == fl->iter_domain() ||
+                gpu_lower->caParallelMap().areMapped(
+                    produce_at_lowered_it, fl->iter_domain());
```

Comment:
nit: no need to check `produce_at_lowered_it == fl->iter_domain()` as identical pointers are always considered mapped.